go/ct/gen/block_context.go

```diff
@@ -68,13 +238,191 @@ func (*BlockContextGenerator) Generate(rnd *rand.Rand, revision common.Revision)
+		b.blockNumberSolver = o.blockNumberSolver.Clone()
+	}
+	b.rangeConstraints = maps.Clone(o.rangeConstraints)
+	b.valueConstraint = maps.Clone(o.valueConstraint)
+}
+
+func (b *BlockContextGenerator) String() string {
+	if b.unsatisfiable {
+		return "false"
+	}
+	if b.blockNumberSolver == nil && b.rangeConstraints == nil && b.valueConstraint == nil {
+		return "true"
+	}
+
+	clauses := []string{}
+	if b.blockNumberSolver != nil {
+		if b.blockNumberSolver.IsSatisfiable() {
+			clauses = append(clauses, fmt.Sprintf(
+				"BlockNumber ∈ [%d..%d]",
+				b.blockNumberSolver.min,
+				b.blockNumberSolver.max,
+			))
+		}
+	}
+
+	if b.rangeConstraints != nil {
+		for variable, inRange := range b.rangeConstraints {
+			if inRange {
+				clauses = append(clauses, fmt.Sprintf(
+					"%v ∈ [BlockNumber-256..BlockNumber-1]",
+					variable,
+				))
+			} else {
+				clauses = append(clauses, fmt.Sprintf(
+					"%v ∉ [BlockNumber-256..BlockNumber-1]",
+					variable,
+				))
+			}
+		}
+	}
+
+	if b.valueConstraint != nil {
+		for variable, value := range b.valueConstraint {
+			op := "+"
+			if value > 0 {
+				op = "-"
+			} else {
+				value = -value
+			}
+			clauses = append(clauses, fmt.Sprintf(
+				"%v = BlockNumber%s%d",
+				variable,
+				op,
+				value,
+			))
+		}
+	}
+
+	sort.Slice(clauses, func(i, j int) bool {
+		return clauses[i] < clauses[j]
+	})
+
+	return strings.Join(clauses, " Λ ")
+}
+
+// RestricVariableToOneOfTheLast256Blocks adds a constraint on the variable
+// so that it is generated with a value within the last 256 blocks.'
+// If the generator is already marked as unsatisfiable or if the variable
+// is already constrained to be outside the last 256 blocks, the generator
+// remains unsatisfiable.
+func (b *BlockContextGenerator) RestrictVariableToOneOfTheLast256Blocks(variable Variable) {
+	if b.unsatisfiable {
+		return
+	}
+	if b.rangeConstraints == nil {
+		b.rangeConstraints = make(map[Variable]bool)
+	}
+	if inRange, ok := b.rangeConstraints[variable]; ok {
+		if !inRange {
+			b.markUnsatisfiable()
+		}
+	} else {
+		b.rangeConstraints[variable] = true
+	}
+}
+
+// RestrictVariableToNoneOfTheLast256Blocks adds a constraint on the variable
+// so that it is generated with a value outside the last 256 blocks.'
+// If the generator is already marked as unsatisfiable or if the variable
+// is already constrained to be within the last 256 blocks, the generator
+// remains unsatisfiable.
+func (b *BlockContextGenerator) RestrictVariableToNoneOfTheLast256Blocks(variable Variable) {
+	if b.unsatisfiable {
+		return
+	}
+	if b.rangeConstraints == nil {
+		b.rangeConstraints = make(map[Variable]bool)
+	}
+	if inRange, ok := b.rangeConstraints[variable]; ok {
+		if inRange {
+			b.markUnsatisfiable()
+		}
+	} else {
+		b.rangeConstraints[variable] = false
+	}
+}
+
+// SetBlockNumberOffsetValue adds a constraint on the variable so that it is
+// generated with a value that is offset from the block number.
```

Comment:
A variable is not "generated". The Generator process assigns a value to it. Please fix the comment.